### PR TITLE
feat: cast actions deeplink v2

### PR DESCRIPTION
**Pull request by dalechyn**
I've set changesets to a minor bump as it is gonna break the API of `Button.AddCastAction` and is gonna require a user to pass the options as the third argument of the `.castAction` handler.

While being a breaking change, it is pretty easy to migrate it.

### CODE REVIEW

File: src/frog-base.tsx [hunk 1]
```diff
@@ -300,15 +314,36 @@ export class FrogBase<
...
+      const origin = this.origin ?? url.origin
+      const baseUrl = origin + parsePath(this.basePath)
+
+      if (c.req.method === 'GET') {
+        console.log({
+          ...installParameters,
+          action: {
+            type: 'post',
+          },
+        })
```
Comment (dalechyn) [inline src/frog-base.tsx]:
```suggestion
```

File: src/types/routes.ts [hunk 2]
```diff
@@ -93,328 +93,753 @@ export type H<
...
-    options?: RouteOptions,
-  ): FrogBase<
-    E,
-    S &
-      ToSchema<M, MergePath<BasePath, P>, I10['in'], MergeTypedResponseData<R>>,
-    BasePath
-  >
-}
+> = /* We enforce `options` parameter to not be partial for Cast Actions*/
+M extends 'cast-action'
```
Comment (jxom) [inline src/types/routes.ts]: Hmm. I would rather avoid this ternary. Can we instead make `options` optional within the parameter definition itself if the method extends `"cast-action"`?

```tsx
(
  path: P,
  handler: H<...>,
  ...rest: (M extends 'cast-action' ? [options: RouteOptions<M>] : [options?: RouteOptions<M>])
): FrogBase<...>
```

File: src/frog-base.tsx [hunk 3]
```diff
@@ -168,9 +169,22 @@ export type FrogConstructorParameters<
   verify?: boolean | 'silent' | undefined
 }
 
-export type RouteOptions = Pick<FrogConstructorParameters, 'verify'> & {
-  fonts?: ImageOptions['fonts'] | (() => Promise<ImageOptions['fonts']>)
-}
+export type RouteOptions<M extends string = string> = Pick<
```
Comment (jxom) [inline src/frog-base.tsx]:
```suggestion
export type RouteOptions<method extends string = string> = Pick<
```

We can conform to our convention of generics as we aren't working on the Hono type fork (`types/routes.ts`) anymore

File: src/frog-base.tsx [hunk 4]
```diff
@@ -168,9 +169,22 @@ export type FrogConstructorParameters<
...
+  FrogConstructorParameters,
+  'verify'
+> &
+  (M extends 'frame'
+    ? {
+        fonts?: ImageOptions['fonts'] | (() => Promise<ImageOptions['fonts']>)
+      }
+    : M extends 'cast-action'
```
Comment (jxom) [inline src/frog-base.tsx]:
```suggestion
    : M extends 'castAction'
```

Would be good if it mapped to the casing of the method.

File: src/utils/getRouteParameters.ts [hunk 5]
```diff
@@ -17,5 +17,5 @@ export function getRouteParameters<env extends Env, handler>(
     else middlewares.push(parameters[i])
   }
 
-  return [parameters[0], middlewares, handler!, options ?? {}]
```
Comment (jxom) [inline src/utils/getRouteParameters.ts:20]: Is this diff intended? Previous implementation is correct, right?

File: src/types/routes.ts (hunk 2 above)
Comment (dalechyn) [reply src/types/routes.ts]: brilliant

File: src/utils/getRouteParameters.ts (hunk 5 above)
Comment (dalechyn) [reply src/utils/getRouteParameters.ts:20]: it was intended, just fixed it with a more narrowed typing.

File: src/frog-base.tsx (hunk 4 above)
Comment (dalechyn) [reply src/frog-base.tsx]: done!

File: src/frog-base.tsx (hunk 3 above)
Comment (dalechyn) [reply src/frog-base.tsx]: done!

File: src/types/routes.ts (hunk 2 above)
Comment (dalechyn) [reply src/types/routes.ts]: done!

**Comment by dalechyn** [comment]
need to correct the docs here

**Review by jxom — APPROVED** [review]

**Pull request merged**